File: java-samples/encumbrance-pawn-shop/workflows/src/main/java/com/r3/developers/samples/encumbrance/workflows/GetLoanFlow.java
package com.r3.developers.samples.encumbrance.workflows;

import com.r3.developers.samples.encumbrance.states.Loan;
import net.corda.v5.application.flows.ClientRequestBody;
import net.corda.v5.application.flows.ClientStartableFlow;
import net.corda.v5.application.flows.CordaInject;
import net.corda.v5.application.marshalling.JsonMarshallingService;
import net.corda.v5.base.annotations.Suspendable;
import net.corda.v5.base.types.MemberX500Name;
import net.corda.v5.ledger.utxo.UtxoLedgerService;
import org.jetbrains.annotations.NotNull;

import java.util.List;
import java.util.stream.Collectors;

public class GetLoanFlow implements ClientStartableFlow {

    @CordaInject
    public JsonMarshallingService jsonMarshallingService;

    @CordaInject
    public UtxoLedgerService ledgerService;

    @NotNull
    @Override
    @Suspendable
    public String call(@NotNull ClientRequestBody requestBody) {
        List<LoanDetail> loanList =
                ledgerService.findUnconsumedStatesByType(Loan.class).stream().map(
                        it -> new LoanDetail(
                                it.getState().getContractState().getLoanId(),
                                it.getState().getContractState().getLender().getName(),
                                it.getState().getContractState().getBorrower().getName(),
                                it.getState().getContractState().getLoanAmount(),
                                it.getState().getContractState().getCollateral(),
                                it.getState().getEncumbranceGroup() != null ?
                                        it.getState().getEncumbranceGroup().getTag() : null,
                                it.getState().getEncumbranceGroup() != null ?
                                        it.getState().getEncumbranceGroup().getSize() : 0
                        )
                ).collect(Collectors.toList());
        return jsonMarshallingService.format(loanList);
    }
}

class LoanDetail {
    private String loanId;
    private MemberX500Name lender;
    private MemberX500Name borrower;
    private int loanAmount;
    private  String collateral;
    private String encumbranceGroupTag;
    private int encumbranceGroupSize;

    public LoanDetail(String loanId, MemberX500Name lender, MemberX500Name borrower, int loanAmount, String collateral,
                      String encumbranceGroupTag, int encumbranceGroupSize) {
        this.loanId = loanId;
        this.lender = lender;
        this.borrower = borrower;
        this.loanAmount = loanAmount;
        this.collateral = collateral;
        this.encumbranceGroupTag = encumbranceGroupTag;
        this.encumbranceGroupSize = encumbranceGroupSize;
    }

    public String getLoanId() {
        return loanId;
    }

    public void setLoanId(String loanId) {
        this.loanId = loanId;
    }

    public MemberX500Name getLender() {
        return lender;
    }

    public void setLender(MemberX500Name lender) {
        this.lender = lender;
    }

    public MemberX500Name getBorrower() {
        return borrower;
    }

    public void setBorrower(MemberX500Name borrower) {
        this.borrower = borrower;
    }

    public int getLoanAmount() {
        return loanAmount;
    }

    public void setLoanAmount(int loanAmount) {
        this.loanAmount = loanAmount;
    }

    public String getEncumbranceGroupTag() {
        return encumbranceGroupTag;
    }

    public void setEncumbranceGroupTag(String encumbranceGroupTag) {
        this.encumbranceGroupTag = encumbranceGroupTag;
    }

    public int getEncumbranceGroupSize() {
        return encumbranceGroupSize;
    }

    public void setEncumbranceGroupSize(int encumbranceGroupSize) {
        this.encumbranceGroupSize = encumbranceGroupSize;
    }

    public String getCollateral() {
        return collateral;
    }

    public void setCollateral(String collateral) {
        this.collateral = collateral;
    }
}

File: kotlin-samples/encumbrance-pawn-shop/workflows/src/main/kotlin/com/r3/developers/samples/encumbrance/workflows/GetLoanFlow.kt
package com.r3.developers.samples.encumbrance.workflows

import com.r3.developers.samples.encumbrance.states.Loan
import net.corda.v5.application.flows.ClientRequestBody
import net.corda.v5.application.flows.ClientStartableFlow
import net.corda.v5.application.flows.CordaInject
import net.corda.v5.application.marshalling.JsonMarshallingService
import net.corda.v5.base.annotations.Suspendable
import net.corda.v5.base.types.MemberX500Name
import net.corda.v5.ledger.utxo.UtxoLedgerService
import org.slf4j.LoggerFactory

data class LoanDetail(val  loanId:String,
                      val  lender: MemberX500Name,
                      val  borrower:MemberX500Name,
                      val  loanAmount: Int,
                      val   collateral:String,
                      val  encumbranceGroupTag:String?,
                      val  encumbranceGroupSize:Int?)


class GetLoanFlow : ClientStartableFlow {

    private companion object {
        val log = LoggerFactory.getLogger(this::class.java.enclosingClass)
    }
    @CordaInject
    lateinit var jsonMarshallingService: JsonMarshallingService

    // Injects the UtxoLedgerService to enable the flow to make use of the Ledger API.
    @CordaInject
    lateinit var ledgerService: UtxoLedgerService

    @Suspendable
    override fun call(requestBody: ClientRequestBody): String {
        log.info("GetLoanFlow.call() called")

        // Queries the VNode's vault for unconsumed states and converts the result to a serializable DTO.
        val states = ledgerService.findUnconsumedStatesByType(Loan::class.java)
        val results = states.map {it ->
            LoanDetail(
                it.state.contractState.loanId,
                it.state.contractState.lender.name,
                it.state.contractState.borrower.name,
                it.state.contractState.loanAmount,
                it.state.contractState.collateral,
                it.state.encumbranceGroup?.tag,
                it.state.encumbranceGroup?.size
            )
        }

        // Uses the JsonMarshallingService's format() function to serialize the DTO to Json.
        return jsonMarshallingService.format(results)
    }
}
/*
  {
     "clientRequestId": "get-loan",
     "flowClassName": "com.r3.developers.samples.encumbrance.workflows.GetLoanFlow",
     "requestBody": {
     }
  }
 */